File: 키패드누르기.py
def find_num(btn,num):
    for x in range(4):
        for y in range(3):
            if btn[x][y] == num:
                numX = x
                numY = y
    return numX, numY
                    
def solution(numbers, hand):
    answer = ''    
    btn = [[1,2,3],[4,5,6],[7,8,9],['*',0,'#']]
    pos_L_X = 3
    pos_L_Y = 0
    pos_R_X = 3
    pos_R_Y = 2
    for num in numbers:
        numX, numY = find_num(btn,num)
        if num == 1 or num == 4 or num == 7:
            answer += 'L'
            pos_L_X = numX
            pos_L_Y = numY
        if num == 3 or num == 6 or num == 9:
            answer += 'R'
            pos_R_X = numX
            pos_R_Y = numY
        if num == 2 or num == 5 or num == 8 or num == 0:
            l = abs(numX-pos_L_X)+abs(numY-pos_L_Y)
            r = abs(numX-pos_R_X)+abs(numY-pos_R_Y)
            if l < r:
                pos_L_X = numX
                pos_L_Y = numY
                answer += "L"
            if l > r:
                pos_R_X = numX
                pos_R_Y = numY
                answer += "R"
            if l == r:
                if hand == "right":
                    pos_R_X = numX
                    pos_R_Y = numY
                    answer += "R"
                else:
                    pos_L_X = numX
                    pos_L_Y = numY
                    answer += "L"
    return answer

File: Programmers/키패드누르기.js
function solution(numbers, hand) {
  function findKey(key) {
    let keypad = [
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
      ["*", 0, "#"],
    ];
    for (let i = 0; i < 4; i++) {
      for (let j = 0; j < 3; j++) {
        if (keypad[i][j] === key) {
          return [i, j];
        }
      }
    }
  }

  let answer = "";
  let left = "*";
  let right = "#";
  for (let i of numbers) {
    if (i === 1 || i === 4 || i === 7) {
      answer += "L";
      left = i;
    } else if (i === 3 || i === 6 || i === 9) {
      answer += "R";
      right = i - 2;
    } else {
      let r = findKey(right);
      let l = findKey(left);
      let middle = findKey(i);
      let rr = Math.abs(r[0] - middle[0]) + Math.abs(r[1] - middle[1]);
      let ll = Math.abs(l[0] - middle[0]) + Math.abs(l[1] - middle[1]);
      if (rr === ll) {
        hand === "right"
          ? ((right = i), (answer += "R"))
          : ((left = i), (answer += "L"));
      } else if (rr > ll) {
        answer += "L";
        left = i;
      } else {
        answer += "R";
        right = i;
      }
    }
  }

  return answer;
}
